text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1546 @@
+    └────────────────┘                  └─────────────────────────────┘
+            │                                          │
+         implies                                    implies
+            │                                          │
+            ↓                                          ↓
+┌───────────────────────┐             ┌────────────────────────────────────┐
+│ const ValueSized      │ ──────────→ │ ValueSized                         │
+│ {type, target, value} │   implies   │ {type, target, runtime env, value} │
+└───────────────────────┘             └────────────────────────────────────┘
+                                                       │
+                                                    implies
+                                                       │
+                               ┌───────────────────────┘
+                               ↓
+                     ┌──────────────────┐
+                     │ Pointee          │
+                     │ {runtime env, *} │
+                     └──────────────────┘
+                               │
+                            implies
+                               │
+                               ↓
+                     ┌──────────────────┐
+                     │ Value            │
+                     │ {runtime env, *} │
+                     └──────────────────┘
+```
+
+`Pointee` is still defined as being implemented for any type that can be used
+behind a pointer and may not be sized at all, this would be implemented for
+effectively every type except wasm's `externref` (or similar opaque types from
+some GPU targets). `Value` is defined as being implemented for any type that can
+be used as a value, which is all types, and also may not be sized at all.
+
+Earlier in this RFC, `extern type`s have previously been described as not being
+able to be used as a value, but it could instead be permitted to write functions
+which use extern types as values (e.g. such as taking an extern type as an argument),
+and instead rely on it being impossible to get a extern type that is not behind a
+pointer or a reference. This also implies that `ValueSized` types can be used as values,
+which would remain prohibited behind the `unsized_locals` and `unsized_fn_params`
+features until these are stabilised.
+
+With these changes to the RFC and possibility additional changes to the language, it
+could be possible to support wasm's `externref` and opaque types from some GPU targets.
+
+## Alignment
+[alignment]: #alignment
+
+There has been community interest in an [`Aligned` trait][rfc_aligned] and there
+are examples of `Aligned` traits being added in the ecosystem:
+
+- `rustc` has [its own `Aligned` trait][rustc_aligned] to support pointer tagging.
+- [`unsized-vec`][crate_unsized_vec] implements a `Vec` that depends on knowing
+  whether a type has an alignment or not.
+
+An `Aligned` trait could be added to this proposal between `ValueSized` and `Pointee`
```

Comment:
I think this is the wrong place to put it, `dyn Trait` is `ValueSized` but is not `Aligned`, but I can imagine extern types that are `Aligned` but not `ValueSized`. I think that means `Aligned` could be a super trait of `Sized` but doesn't otherwise fit in this hierarchy. (In fact I think there's an entirely separate hierarchy including `ValueAligned`, etc.)

Reply:
I've rewritten this section of the RFC.